### PR TITLE
[mclag]:add mclagsyncd

### PR DESCRIPTION
**Pull request by shine4chen**
**What I did**
add mclagsyncd

**Why I did it**
accomplish it according to MCLAG HLD document

**How I verified it**
test it on nephos lab

**Details if related**
Please refer #https://github.com/Azure/SONiC/pull/325

### CODE REVIEW

File: mclagsyncd/mclaglink.cpp [hunk 1]
```diff
@@ -0,0 +1,814 @@
...
+void MclagLink::mclag_connection_lost_handle_port_isolate()
+{
+    string acl_name = "mclag";
+
+    p_acl_tbl->del(acl_name);
+
+    return;
+}
+
+void MclagLink::mclag_connection_lost_handle_port_learn_mode()
```
Comment (lisahnguyen) [inline mclagsyncd/mclaglink.cpp]: What is the intended usage of this API and the MclagLink::mclag_connection_lost_handle_port_isolate()? Do you plan to revert all changes to hardware when mclagsyncd's connection to ICCPd fails? If yes, please elaborate more on the overall scheme and traffic impact implication.

File: mclagsyncd/mclaglink.cpp [hunk 2]
```diff
@@ -0,0 +1,814 @@
...
+            msg_head->msg_len = (unsigned short)infor_len;
+            msg_head->msg_type = MCLAG_SYNCD_MSG_TYPE_FDB_OPERATION;
+
+            SWSS_LOG_NOTICE("mclagsycnd send msg to iccpd, msg_len =%d, msg_type =%d",
+                            msg_head->msg_len, msg_head->msg_type);
+            write = ::write(m_connection_socket, infor_start, msg_head->msg_len);
+            if (write <= 0)
+                return write;
+
+            infor_len = 0;
```
Comment (gitsabari) [inline mclagsyncd/mclaglink.cpp]: i believe infor_len should be initialised to header size. infor_len=sizeof(mclag_msg_hdr_t); same issue in line 580 also

File: mclagsyncd/mclaglink.cpp (hunk 1 above)
Comment (jianjundong) [reply mclagsyncd/mclaglink.cpp]: @lisahnguyen, We will delete the unused codes.

File: mclagsyncd/mclaglink.cpp (hunk 2 above)
Comment (jianjundong) [reply mclagsyncd/mclaglink.cpp]: @gitsabari, We will modify it ASAP. Thanks.

File: mclagsyncd/mclaglink.h [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+#include "producerstatetable.h"
+#include "selectable.h"
+#include "redisclient.h"
+#include "mclagsyncd/mclag.h"
+
+namespace swss {
+
+struct mclag_fdb_info
+{
+    char mac[32];
```
[the remark this answers is not in the document]
Comment (jianjundong) [reply mclagsyncd/mclaglink.h]: @gitsabari, This is string format, we will change this string from 32 bytes to 18 bytes.

File: mclagsyncd/mclaglink.h [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+class MclagLink : public Selectable {
+public:
+    const int MSG_BATCH_SIZE;
+    std::map<std::string, std:: string> *p_learn;
+    ProducerStateTable * p_port_tbl;
+    ProducerStateTable * p_lag_tbl;
+    ProducerStateTable * p_tnl_tbl;
+    ProducerStateTable * p_intf_tbl;
+    ProducerStateTable *p_fdb_tbl;
+    ProducerStateTable *p_acl_tbl;
```
Comment (stcheng) [inline mclagsyncd/mclaglink.h]: change to `p_acl_table_tbl`

File: mclagsyncd/mclaglink.h [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+    ProducerStateTable *p_acl_rule_tbl;
+    DBConnector *p_appl_db;
+    RedisClient *p_redisClient_2_asic;
```
Comment (stcheng) [inline mclagsyncd/mclaglink.h]: what does the variable's name mean?

File: mclagsyncd/mclaglink.cpp [hunk 6]
```diff
@@ -0,0 +1,814 @@
...
+#include "swss/notificationproducer.h"
+#include "mclagsyncd/mclaglink.h"
+#include "mclagsyncd/mclag.h"
+#include <set>
+#include <algorithm>
+
+using namespace swss;
+using namespace std;
+
+vector<string> string_split(const string& s, const string& delim)
```
Comment (stcheng) [inline mclagsyncd/mclaglink.cpp]: please use the `tokenize` function in swss-common repository

File: mclagsyncd/mclaglink.h [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+    char *m_messageBuffer;
+    char *m_messageBuffer_send;
+    unsigned int m_pos;
+
+    bool m_connected;
+    bool m_server_up;
+    int m_server_socket;
+    int m_connection_socket;
+
+    void mclagsyncd_get_oid_2_port_name_map(std::unordered_map<std::string, std:: string> & port_map);
```
Comment (stcheng) [inline mclagsyncd/mclaglink.h]: could you remove prefix `mclagsyncd`? and also replace the `2` with `to`? and in swss repository, all the function names are camelCased.

File: mclagsyncd/mclaglink.cpp [hunk 8]
```diff
@@ -0,0 +1,814 @@
...
+        }
+
+        if (memcmp(type_it->second.c_str(), "SAI_FDB_ENTRY_TYPE_DYNAMIC", type_it->second.length()) == 0)
+            type = "dynamic";
+        else
+            type = "static";
+
+        /*get port name*/
+        mclagsyncd_get_oid_2_port_name_map(oid_2_portname_map);
+#if 0
```
Comment (stcheng) [inline mclagsyncd/mclaglink.cpp]: you could change it to DEBUG level

File: mclagsyncd/mclaglink.cpp [hunk 9]
```diff
@@ -0,0 +1,814 @@
...
+    isolate_src_port.insert(0, (const char*)cur, op_hdr->op_len);
+
+    cur = cur + op_hdr->op_len;
+
+    /*get isolate dst ports infor*/
+    op_hdr = (mclag_sub_option_hdr_t *)cur;
+    cur = cur + MCLAG_SUB_OPTION_HDR_LEN;
+    isolate_dst_port.insert(0, (const char*)cur, op_hdr->op_len);
+
+    SWSS_LOG_NOTICE("Enter mclagsyncd_set_port_isolate");
```
Comment (stcheng) [inline mclagsyncd/mclaglink.cpp]: you shall use `SWSS_LOG_ENTER`

Comment (shine4chen) [reply mclagsyncd/mclaglink.cpp]: Sure , we will remove some unnecessary log.

File: mclagsyncd/mclaglink.h (hunk 7 above)
Comment (shine4chen) [reply mclagsyncd/mclaglink.h]: Sure, done

File: mclagsyncd/mclaglink.cpp (hunk 6 above)
Comment (shine4chen) [reply mclagsyncd/mclaglink.cpp]: Sure, done

File: mclagsyncd/mclaglink.h (hunk 4 above)
Comment (shine4chen) [reply mclagsyncd/mclaglink.h]: Sure ,done

File: mclagsyncd/mclaglink.cpp (hunk 8 above)
Comment (shine4chen) [reply mclagsyncd/mclaglink.cpp]: Sure, done

File: mclagsyncd/mclaglink.h (hunk 5 above)
Comment (shine4chen) [reply mclagsyncd/mclaglink.h]: we has changed it to `p_redisClient_to_asic`. It is used to access asic_db.

**Comment by shine4chen** [comment]
retest this please.

**Comment by shine4chen** [comment]
retest this please

**Comment by lguohan** [comment]
I saw binary in the pr. please remove the binary.

**Comment by shine4chen** [comment]
@lguohan Thanks, done

**Comment by shine4chen** [comment]
retest this please

**Comment by shine4chen** [comment]
retest this please

**Comment by shine4chen** [comment]
retest this please

**Comment by shine4chen** [comment]
@stcheng @rlhui @gitsabari @lisahnguyen I resolved some conflict and system automatically dismiss your approve. Would you please help to review and approve it again ?

**Pull request merged**